### PR TITLE
[MetaSchedule] Introduce MMA Tensor Core Multilevel Tiling

### PR DESCRIPTION
**Pull request by cblmemo**
This PR introduces mma in the multilevel tiling tensor core.

For the benchmark result, please refer to https://docs.google.com/spreadsheets/d/1thf1jsbX87WokRfESXO14fx40H3vYHDk6EWkb_wnv5Y

For all tuning logs, best performance scripts and python tuning & benchmarking scripts, please refer to https://github.com/cblmemo/TVMGemmAsync/tree/main/mma

### CODE REVIEW

File: python/tvm/tir/tensor_intrin/cuda.py [hunk 1]
```diff
@@ -1182,3 +1182,263 @@ def get_wmma_intrin_group(
...
+TensorIntrin.register(
+    "m16n8k8_load_A_row_major", m16n8k8_load_A_row_major_desc, m16n8k8_load_A_row_major_impl
+)
+TensorIntrin.register(
+    "m16n8k8_load_B_row_major", m16n8k8_load_B_row_major_desc, m16n8k8_load_B_row_major_impl
+)
+TensorIntrin.register("m16n8k8_sync", m16n8k8_sync_desc, m16n8k8_sync_impl)
+TensorIntrin.register(
+    "m16n8k8_store_C_row_major", m16n8k8_store_C_row_major_desc, m16n8k8_store_C_row_major_impl
+)
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/cuda.py:1438]: Why can't the existing intrinsic definitions for m16n16k16 be used?

**Comment by FrozenGene** [comment]
How to reproduce it? I would like to test it on A100 @cblmemo

**Comment by cblmemo** [comment]
> How to reproduce it? I would like to test it on A100 @cblmemo

You could find all scripts you need at https://github.com/cblmemo/TVMGemmAsync/tree/main/mma 🫡

**Comment by FrozenGene** [comment]
> > How to reproduce it? I would like to test it on A100 @cblmemo
> 
> You could find all scripts you need at https://github.com/cblmemo/TVMGemmAsync/tree/main/mma 🫡 @FrozenGene

@cblmemo Small bug: https://github.com/cblmemo/TVMGemmAsync/blob/main/mma/GemmRuleGenerate.py#L135 should be `tensorcore_outputs` not `outputs`

**Comment by FrozenGene** [comment]
@cblmemo Great job! I have tested it on A100. 1024x1024x1024, it could achieve the same level of cutlass. If we could add more mma variants, maybe we could achieve better result.

**Comment by cblmemo** [comment]
> > > How to reproduce it? I would like to test it on A100 @cblmemo
> > 
> > 
> > You could find all scripts you need at https://github.com/cblmemo/TVMGemmAsync/tree/main/mma 🫡 @FrozenGene
> 
> @cblmemo Small bug: https://github.com/cblmemo/TVMGemmAsync/blob/main/mma/GemmRuleGenerate.py#L135 should be `tensorcore_outputs` not `outputs`

@FrozenGene Thanks for point that out! I use a lot of dir names and it appears than I made a mistake when uploading my script 🙂

**Comment by cblmemo** [comment]
> @cblmemo Great job! I have tested it on A100. 1024x1024x1024, it could achieve the same level of cutlass. If we could add more mma variants, maybe we could achieve better result.

@FrozenGene Sure. The m16n8k16 and fp32 accumulator are WIP now 🧐

**Comment by FrozenGene** [comment]
> > @cblmemo Great job! I have tested it on A100. 1024x1024x1024, it could achieve the same level of cutlass. If we could add more mma variants, maybe we could achieve better result.
> 
> @FrozenGene Sure. The m16n8k16 and fp32 accumulator are WIP now 🧐

@cblmemo Sounds great! Also consider supporting signed (and unsigned ) i8 * i8 -> i32 (accumulator)(m16n8k32/m8n8k16), which is commonly used in the quantized model. If we have this, we could do more interesting benchmark!

File: include/tvm/tir/transform.h [hunk 2]
```diff
@@ -481,7 +481,7 @@ TVM_DLL Pass ConvertBlocksToOpaque();
  * Before narrowing, `B` is a `[16, 16]` buffer, but only a skinny vector `B[i, 0:16]` is accessed.
  *
  *  \code
- *
+ *∏
```
Comment (Hzfengsy) [inline include/tvm/tir/transform.h:484]: typo here

File: python/tvm/tir/schedule/schedule.py [hunk 3]
```diff
@@ -402,6 +402,46 @@ def sample_perfect_tile(
...
+            The factor of the second part
+        decision: Optional[List[int]]
+            The sampling decision, if any
+
+        Returns
+        -------
+        result : List[ExprRV]
+            A list of length `n`, the random partitioned tile sizes sampled
+        """
+        return list(
```
Comment (Hzfengsy) [inline python/tvm/tir/schedule/schedule.py:434]: Not sure if explicit conversion is necessary

File: python/tvm/tir/tensor_intrin/cuda.py [hunk 4]
```diff
@@ -530,17 +530,17 @@ def mma_store_impl(a: T.handle, c: T.handle) -> None:
 
 MMA_store_16x16_f32_global_INTRIN = "mma_store_16x16_f32_global_"
 TensorIntrin.register(
-    MMA_store_16x16_f32_global_INTRIN, *get_mma_store_intrin("float32", 8, "global")
+    MMA_store_16x16_f32_global_INTRIN, *get_mma_stmatrix_intrin("float32", 8, "global")
```
Comment (Hzfengsy) [inline python/tvm/tir/tensor_intrin/cuda.py:533]: Any reason to rename it?

File: src/driver/driver_api.cc [hunk 5]
```diff
@@ -202,14 +202,18 @@ Array<tvm::transform::Pass> CreatePassList(bool disable_loop_partition) {
...
+  pass_list.push_back(tir::transform::UnifyThreadBinding());
   pass_list.push_back(tir::transform::LowerMatchBuffer());
+  pass_list.push_back(tir::transform::Simplify());
+  pass_list.push_back(tir::transform::InjectPermutedLayout());
+  pass_list.push_back(tir::transform::Simplify());
   pass_list.push_back(tir::transform::InjectSoftwarePipeline());
-  pass_list.push_back(tir::transform::LowerOpaqueBlock());
+  pass_list.push_back(tir::transform::TransformMmaBufferLayout());
   pass_list.push_back(tir::transform::FlattenBuffer());
+  pass_list.push_back(tir::transform::LowerOpaqueBlock());
```
Comment (Hzfengsy) [inline src/driver/driver_api.cc:216]: Why `LowerOpaqueBlock` should be called after `FlattenBuffer`? I know it needs to be after the `TransformMmaBufferLayout`, but could it be before `FlattenBuffer`?

Comment (cblmemo) [reply src/driver/driver_api.cc:215]: Yes. Now the order is correct

File: python/tvm/tir/schedule/schedule.py (hunk 3 above)
Comment (cblmemo) [reply python/tvm/tir/schedule/schedule.py:434]: The `sample_perfect_tile` primitive also use this explicit conversion

File: python/tvm/tir/tensor_intrin/cuda.py (hunk 4 above)
Comment (cblmemo) [reply python/tvm/tir/tensor_intrin/cuda.py]: It was changed back now

File: src/tir/transforms/inject_permuted_layout.cc [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        store = if_then_else->then_case.as<BufferStoreNode>();
+        ICHECK(!if_then_else->else_case);
+        if_then_else_condition = if_then_else->condition;
+      }
+      ICHECK(store) << body;
+
+      // Step 1.3. Get smem width and refuse to make any difference if invalid
+      auto smem_width = store->buffer->shape[1].as<IntImmNode>()->value;
+      if (smem_width % 32 != 0) {
+        LOG(INFO) << "Permuted Layout for " << op->block->name_hint
```
Comment (Hzfengsy) [inline src/tir/transforms/inject_permuted_layout.cc:145]:
```suggestion
        LOG(WARNING) << "Permuted Layout for " << op->block->name_hint
```

File: src/tir/transforms/inject_permuted_layout.cc [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                  << " is not supported since its second dimension is not divisible by 32";
+        return br;
+      }
+      if (smem_width % 64 == 32) {
+        if (store->buffer->shape[0].as<IntImmNode>()->value % 2 != 0) {
+          LOG(INFO) << "Permuted Layout for " << op->block->name_hint
```
Comment (Hzfengsy) [inline src/tir/transforms/inject_permuted_layout.cc:151]:
```suggestion
          LOG(WARNING) << "Permuted Layout for " << op->block->name_hint
```

**Review by Hzfengsy — APPROVED** [review]
LGTM

**Review by spectrometerHBH — APPROVED** [review]

**Review by FrozenGene — APPROVED** [review]

**Comment by Hzfengsy** [comment]
Thanks @cblmemo for the excellent work, together with the reviews from @spectrometerHBH and @FrozenGene

**Pull request merged**